File: offer/18/JPhoebe.py
# -*- coding:utf-8 -*-

# 题目描述
# 操作给定的二叉树，将其变换为源二叉树的镜像。
# 输入描述:
# 二叉树的镜像定义：源二叉树
# 8
# / \
#     6   10
# / \  / \
#     5  7 9 11
# 镜像二叉树
# 8
# / \
#     10   6
# / \  / \
#     11 9 7  5

class Solution:
    def Mirror(self, root):
        # 递归更换左右节点
        self.hand_mirror(root)
        return root

    def hand_mirror(self, root):
        if root is None:
            return
        temp = root.left
        root.left = root.right
        root.right = temp
        self.hand_mirror(root.left)
        self.hand_mirror(root.right)

File: offer/61/jphoebe.java
/*
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
public class Solution {
    StringBuffer result = new StringBuffer();
    int index = -1;
    String[] strSplit = null;

    String Serialize(TreeNode root) {
        if (root == null){
            return result.append("#!").toString();
        }
        result.append(root.val).append("!");
        this.Serialize(root.left);
        this.Serialize(root.right);
        return result.toString();
    }

    TreeNode Deserialize(String str) {
        strSplit = str.split("!");
        TreeNode node = null;
        index++;
        if(!strSplit[index].equals("#")) {
            node = new TreeNode(Integer.valueOf(strSplit[index]));
            node.left = Deserialize(str);
            node.right = Deserialize(str);

        }
        return node;
    }
}
